_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,782 @@
+
+Now let's dive in and see how we may accomplish our goal with `tfprobability`. We start with the simulated dataset.
+
+## Uncertainty estimates on simulated data
+
+### Dataset
+
+
+We re-use the dataset from the Google TensorFlow Probability team's [blog post on the same subject]( https://medium.com/tensorflow/regression-with-probabilistic-layers-in-tensorflow-probability-e46ff5d37baf) ^[see also the corresponding [notebook](https://colab.research.google.com/github/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Probabilistic_Layers_Regression.ipynb#scrollTo=5zCEYpzu7bDX)], with one exception: We extend the range of the independent variable a bit on the negative side, to better demonstrate the different methods' behaviors.
+
+Here is the data-generating process. We also get library loading out of the way. Like the preceding posts on `tfprobability`, this one too features recently added functionality, so please use the development versions of `tensorflow` and `tfprobability` as well as `keras`. Call `install_tensorflow(version = "nightly")` to obtain a current nightly build of TensorFlow and TensorFlow Probability:
+
+
+```{r}
+# make sure we use the development versions of tensorflow, tfprobability and keras
+devtools::install_github("rstudio/tensorflow")
+devtools::install_github("rstudio/tfprobability")
+devtools::install_github("rstudio/keras")
+
+# and that we use a nightly build of TensorFlow and TensorFlow Probability
+tensorflow::install_tensorflow(version = "nightly")
+
+library(tensorflow)
+library(tfprobability)
+library(keras)
+
+library(dplyr)
+library(tidyr)
+library(ggplot2)
+
+# make sure this code is compatible with TensorFlow 2.0
+tf$compat$v1$enable_v2_behavior()
+
+# generate the data
+x_min <- -40
+x_max <- 60
+n <- 150
+w0 <- 0.125
+b0 <- 5
+
+normalize <- function(x) (x - x_min) / (x_max - x_min)
+
+# training data; predictor 
+x <- x_min + (x_max - x_min) * runif(n) %>% as.matrix()
+
+# training data; target
+eps <- rnorm(n) * (3 * (0.25 + (normalize(x)) ^ 2))
+y <- (w0 * x * (1 + sin(x)) + b0) + eps
+
+# test data (predictor)
+x_test <- seq(x_min, x_max, length.out = n) %>% as.matrix()
+```
+
+
+How does the data look?
+
+```{r}
+ggplot(data.frame(x = x, y = y), aes(x, y)) + geom_point()
+```
+
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Simulated data"}
+knitr::include_graphics("images/simdata.png")
+```
+
+
+The task here is single-predictor regression, which in principle we can achieve use Keras `dense` layers.
+Let's see how to enhance this by indicating uncertainty, starting from the aleatoric type.
```

Comment:
```suggestion
Let's see how to enhance this by indicating uncertainty, starting with the aleatoric type.
```

Reply:
^^ I guess this could be either, but my Sprachgefühl pings with. Feel free to ignore!